CircleCI test error: Entrez calls failing

#### Summary:

CI tests are failing due to connection issues against NCBI/Entrez

#### Description:

CircleCI's routine testing threw an error with an Entrez call. This appears to be relevant to #17

```python
============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /home/circleci/repo/venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/circleci/repo
plugins: cov-2.12.1
collecting ... collecting 3 items                                                             collected 7 items                                                              

tests/test_cli_parsing.py::test_bad_infile PASSED                        [ 14%]
tests/test_cli_parsing.py::test_create_and_keep_cache FAILED             [ 28%]
tests/test_cli_parsing.py::test_download_and_log PASSED                  [ 42%]
tests/test_ncfp.py::test_basic_ncbi PASSED                               [ 57%]
tests/test_ncfp.py::test_basic_uniprot PASSED                            [ 71%]
tests/test_ncfp.py::test_basic_stockholm SKIPPED (Database caching n...) [ 85%]
tests/test_ncfp.py::test_small_stockholm PASSED                          [100%]
```

```python
------------------------------ Captured log call -------------------------------
WARNING  ncbi_cds_from_protein.entrez:entrez.py:329 ELing query (XP_005274161.1) failed (retry 1)
Traceback (most recent call last):
  File "/home/circleci/repo/ncbi_cds_from_protein/entrez.py", line 324, in elink_fetch_with_retries
    Entrez.elink(dbfrom=dbname, linkname=linkdbname, id=query_id)
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/Bio/Entrez/__init__.py", line 281, in elink
    return _open(cgi, variables)
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/Bio/Entrez/__init__.py", line 606, in _open
    handle = urlopen(cgi)
  File "/usr/local/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/local/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/local/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/local/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/local/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 429: Too Many Requests

----------- coverage: platform linux, python 3.8.5-final-0 -----------
Coverage XML written to file .coverage.xml

=========================== short test summary info ============================
FAILED tests/test_cli_parsing.py::test_create_and_keep_cache - ncbi_cds_from_...
============== 1 failed, 5 passed, 1 skipped in 78.42s (0:01:18) ===============

Exited with code exit status 1

CircleCI received exit code 1
```

#### Expected Output:

Tests should pass.

#### `ncfp` Version:

0.2.0

#### Python Version:

3.6, 3.7, 3.8

#### Operating System:

Linux (at CircleCI)

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -11,6 +11,7 @@ workflows:
   version: 2.1
   test:
     jobs:
+      - test-3_9
       - test-3_8
       - test-3_7
       - test-3_6
@@ -23,6 +24,7 @@ workflows:
               only:
                 - master
     jobs:
+      - test-3_9
       - test-3_8
       - test-3_7
       - test-3_6
@@ -30,7 +32,7 @@ workflows:
 jobs:
   test-3_8: &test-template
     docker:
-      - image: circleci/python:3.8.5
+      - image: circleci/python:3.8.9
 
     working_directory: ~/repo
 
@@ -82,8 +84,12 @@ jobs:
   test-3_6:
     <<: *test-template
     docker:
-      - image: circleci/python:3.6.8
+      - image: circleci/python:3.6.9
   test-3_7:
     <<: *test-template
     docker:
-      - image: circleci/python:3.7.8
+      - image: circleci/python:3.7.10
+  test-3_9:
+    <<: *test-template
+    docker:
+      - image: circleci/python:3.9.5

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,14 +1,14 @@
 repos:
 -   repo: https://github.com/ambv/black
-    rev: stable
+    rev: 21.5b1
     hooks:
     - id: black
-      language_version: python3.7
+      language_version: python3
 -   repo: https://github.com/asottile/blacken-docs
-    rev: v0.3.0
+    rev: v1.10.0
     hooks:
     -   id: blacken-docs
-        additional_dependencies: [black==18.9b0]
+        additional_dependencies: [black]
 -   repo: https://github.com/pre-commit/pre-commit-hooks
     rev: v1.2.3
     hooks:

--- a/ncbi_cds_from_protein/entrez.py
+++ b/ncbi_cds_from_protein/entrez.py
@@ -314,8 +314,6 @@ def elink_fetch_with_retries(query_id, dbname, linkdbname, maxretries):
     linkdbname      - NCBI target database name for link query
     maxretries      - maximum number of attempts to make
     """
-    import json
-
     logger = logging.getLogger(__name__)
     logger.debug("ELink query: %s", query_id)
 

--- a/ncbi_cds_from_protein/entrez.py
+++ b/ncbi_cds_from_protein/entrez.py
@@ -314,6 +314,8 @@ def elink_fetch_with_retries(query_id, dbname, linkdbname, maxretries):
     linkdbname      - NCBI target database name for link query
     maxretries      - maximum number of attempts to make
     """
+    import json
+
     logger = logging.getLogger(__name__)
     logger.debug("ELink query: %s", query_id)
 
@@ -327,7 +329,7 @@ def elink_fetch_with_retries(query_id, dbname, linkdbname, maxretries):
         except Exception:
             tries += 1
             logger.warning(
-                "ELing query (%s) failed (retry %d)", query_id, tries, exc_info=True
+                "ELink query (%s) failed (retry %d)", query_id, tries, exc_info=True
             )
     raise NCFPMaxretryException("Query ID %s ELink failed" % query_id)
 

--- a/tests/test_cli_parsing.py
+++ b/tests/test_cli_parsing.py
@@ -45,7 +45,6 @@ import pytest
 from unittest.mock import Mock
 
 from Bio import Entrez
-from Bio.Entrez import Parser
 from bioservices import UniProt
 
 from ncbi_cds_from_protein.scripts import ncfp
@@ -127,10 +126,50 @@ def mock_entrez_create_and_keep_cache(monkeypatch):
     responses.side_effect = [
         io.BytesIO(_)
         for _ in (
-            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>10835069</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>197116381</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
-            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>283135214</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>1677498684</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
-            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>530397002</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>767968522</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
-            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>283135242</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>1677500256</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
+            (
+                b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult '
+                b'PUBLIC "-//NLM//DTD elink 20101123//EN" '
+                b'"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  '
+                b"<LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>10835069</Id>\n    "
+                b"</IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      "
+                b"<LinkName>protein_nuccore</LinkName>\n      \n        "
+                b"<Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n        "
+                b"<Link>\n\t\t\t\t<Id>197116381</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  "
+                b"</LinkSet>\n</eLinkResult>\n"
+            ),
+            (
+                b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult '
+                b'PUBLIC "-//NLM//DTD elink 20101123//EN" '
+                b'"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  '
+                b"<LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>283135214</Id>\n    "
+                b"</IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      "
+                b"<LinkName>protein_nuccore</LinkName>\n      \n        "
+                b"<Link>\n\t\t\t\t<Id>1677498684</Id>\n\t\t\t</Link>\n        "
+                b"<Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  "
+                b"</LinkSet>\n</eLinkResult>\n"
+            ),
+            (
+                b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult '
+                b'PUBLIC "-//NLM//DTD elink 20101123//EN" '
+                b'"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  '
+                b"<LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>530397002</Id>\n    "
+                b"</IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      "
+                b"<LinkName>protein_nuccore</LinkName>\n      \n        "
+                b"<Link>\n\t\t\t\t<Id>767968522</Id>\n\t\t\t</Link>\n        "
+                b"<Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  "
+                b"</LinkSet>\n</eLinkResult>\n"
+            ),
+            (
+                b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult '
+                b'PUBLIC "-//NLM//DTD elink 20101123//EN" '
+                b'"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  '
+                b"<LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>283135242</Id>\n    "
+                b"</IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      "
+                b"<LinkName>protein_nuccore</LinkName>\n      \n        "
+                b"<Link>\n\t\t\t\t<Id>1677500256</Id>\n\t\t\t</Link>\n        "
+                b"<Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    "
+                b"</LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n"
+            ),
         )
     ]
 

--- a/tests/test_cli_parsing.py
+++ b/tests/test_cli_parsing.py
@@ -41,6 +41,8 @@
 
 import pytest
 
+from bioservices import UniProt
+
 from ncbi_cds_from_protein.scripts import ncfp
 from ncbi_cds_from_protein import NCFPException
 
@@ -85,6 +87,26 @@ def args_validate_and_log(
     ]
 
 
+@pytest.fixture
+def mock_uniprot_download_and_log(monkeypatch):
+    """Mock remote service call to UniProt for download_and_log test.
+
+    Returns the result expected from l125 of sequences.py, for each query of
+    the path_uniprot_stockholm_small dataset
+
+    u_service.search(match.group(0), columns="database(EMBL)")  # type: ignore
+    """
+
+    def mock_search(*args, **kwargs):
+        """Mock call to UniProt.search() method.
+
+        This output specific to the download_and_log() test
+        """
+        return "Cross-reference (EMBL)\nCM000618;\n"
+
+    monkeypatch.setattr(UniProt, "search", mock_search)
+
+
 def test_bad_infile(args_bad_infile):
     """ncfp stops if CLI input file does not exist.
 
@@ -106,9 +128,10 @@ def test_create_and_keep_cache(args_create_reuse_cache):
     ncfp.run_main(args_create_reuse_cache + ["--keepcache"])
 
 
-def test_download_and_log(args_validate_and_log):
+def test_download_and_log(args_validate_and_log, mock_uniprot_download_and_log):
     """ncfp downloads coding sequences and logs output from CLI.
 
-    Validates that ncfp runs without error, does not check output
+    Validates that ncfp runs without error, does not check output. Calls to UniProt
+    are mocked for use with cloud CI.
     """
     ncfp.run_main(args_validate_and_log)

--- a/tests/test_cli_parsing.py
+++ b/tests/test_cli_parsing.py
@@ -38,9 +38,14 @@
 # THE SOFTWARE.
 """Test command-line parsing for ncfp program"""
 
+import io
 
 import pytest
 
+from unittest.mock import Mock
+
+from Bio import Entrez
+from Bio.Entrez import Parser
 from bioservices import UniProt
 
 from ncbi_cds_from_protein.scripts import ncfp
@@ -65,12 +70,15 @@ def args_create_reuse_cache(email_address, path_human, tmp_path):
         tmp_path / "cache",
         "-c",
         "humancache",
+        "--debug",
     ]
 
 
 @pytest.fixture
 def args_validate_and_log(
-    email_address, path_uniprot_stockholm_small, tmp_path,
+    email_address,
+    path_uniprot_stockholm_small,
+    tmp_path,
 ):
     """Cmd-line arguments for downloading sequences and logging output.
 
@@ -107,6 +115,28 @@ def mock_uniprot_download_and_log(monkeypatch):
     monkeypatch.setattr(UniProt, "search", mock_search)
 
 
+@pytest.fixture
+def mock_entrez_create_and_keep_cache(monkeypatch):
+    """Mock remote service call to NCBI/Entrez for create_and_keep_cache test.
+
+    Returns the result expected from NCBI with a call for each of the proteins
+    in fixtures/sequences/human.fasta
+    """
+
+    responses = Mock()
+    responses.side_effect = [
+        io.BytesIO(_)
+        for _ in (
+            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>10835069</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>197116381</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
+            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>283135214</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>1677498684</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
+            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>530397002</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>767968522</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
+            b'<?xml version="1.0" encoding="UTF-8" ?>\n<!DOCTYPE eLinkResult PUBLIC "-//NLM//DTD elink 20101123//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20101123/elink.dtd">\n<eLinkResult>\n\n  <LinkSet>\n    <DbFrom>protein</DbFrom>\n    <IdList>\n      <Id>283135242</Id>\n    </IdList>\n    <LinkSetDb>\n      <DbTo>nuccore</DbTo>\n      <LinkName>protein_nuccore</LinkName>\n      \n        <Link>\n\t\t\t\t<Id>1677500256</Id>\n\t\t\t</Link>\n        <Link>\n\t\t\t\t<Id>568815587</Id>\n\t\t\t</Link>\n      \n    </LinkSetDb>\n  </LinkSet>\n</eLinkResult>\n',
+        )
+    ]
+
+    monkeypatch.setattr(Entrez, "elink", responses)
+
+
 def test_bad_infile(args_bad_infile):
     """ncfp stops if CLI input file does not exist.
 
@@ -116,7 +146,9 @@ def test_bad_infile(args_bad_infile):
         ncfp.run_main(args_bad_infile)
 
 
-def test_create_and_keep_cache(args_create_reuse_cache):
+def test_create_and_keep_cache(
+    args_create_reuse_cache, mock_entrez_create_and_keep_cache
+):
     """ncfp creates named cache from CLI and keeps it when rerunning.
 
     This calls ncfp twice and expects both calls run without error. Output

--- a/tests/test_ncfp.py
+++ b/tests/test_ncfp.py
@@ -44,8 +44,63 @@ from argparse import Namespace
 
 import pytest
 
+from bioservices import UniProt
+
 from ncbi_cds_from_protein.scripts import ncfp
 from utils import check_files, modify_namespace
+
+
+@pytest.fixture
+def mock_uniprot_download_and_log(monkeypatch):
+    """Mock remote service call to UniProt for download_and_log test.
+
+    Returns the result expected from l125 of sequences.py, for each query of
+    the path_uniprot_stockholm_small dataset
+
+    u_service.search(match.group(0), columns="database(EMBL)")  # type: ignore
+    """
+
+    def mock_search(*args, **kwargs):
+        """Mock call to UniProt.search() method.
+
+        This output specific to the download_and_log() test
+        """
+        return "Cross-reference (EMBL)\nCM000618;\n"
+
+    monkeypatch.setattr(UniProt, "search", mock_search)
+
+
+@pytest.fixture
+def mock_basic_uniprot(monkeypatch):
+    """Mock remote service call to UniProt for test_basic_uniprot test.
+
+    Returns the result expected from l125 of sequences.py, for each query of
+    the path_uniprot_stockholm_small dataset
+
+    u_service.search(match.group(0), columns="database(EMBL)")  # type: ignore
+    """
+    results = iter(
+        [
+            "Cross-reference (EMBL)\nJNBS01004944;\n",
+            "Cross-reference (EMBL)\nJNBS01000225;\n",
+            "Cross-reference (EMBL)\nAZIL01000691;\n",
+            "Cross-reference (EMBL)\nGL833138;\n",
+            "Cross-reference (EMBL)\nJNBR01001477;\n",
+            "Cross-reference (EMBL)\nJNBS01001796;\n",
+            "Cross-reference (EMBL)\nJNBS01000295;\n",
+            "Cross-reference (EMBL)\nKI913977;\n",
+            "Cross-reference (EMBL)\nFN648069;\n",
+        ]
+    )
+
+    def mock_search(*args, **kwargs):
+        """Mock call to UniProt.search() method.
+
+        This output specific to the test_basic_uniprot() test
+        """
+        return next(results)
+
+    monkeypatch.setattr(UniProt, "search", mock_search)
 
 
 @pytest.fixture
@@ -85,7 +140,9 @@ def test_basic_ncbi(namespace_base, path_ncbi, path_ncbi_targets, tmp_path):
     check_files(outdir, path_ncbi_targets, ("ncfp_aa.fasta", "ncfp_nt.fasta"))
 
 
-def test_basic_uniprot(namespace_base, path_uniprot, path_uniprot_targets, tmp_path):
+def test_basic_uniprot(
+    namespace_base, path_uniprot, path_uniprot_targets, tmp_path, mock_basic_uniprot
+):
     """ncfp collects correct coding sequences for basic UniProt input."""
     # Modify default arguments
     infile = path_uniprot

